pkg/gpu/cuda/elf.go

```diff
@@ -0,0 +1,376 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2025-present Datadog, Inc.
+
+package cuda
+
+import (
+	//nolint:depguard
+	// we need the original elf package for all the types/consts
+	"debug/elf"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"iter"
+	"unsafe"
+)
+
+// stringBufferSize is the size of the reusable buffer used to read strings from ELF sections
+const stringBufferSize = 256
+
+// errorUnsupported is returned when an unsupported section is found, which indicates we should skip that section
+var errorUnsupported = fmt.Errorf("unsupported section parsing")
+
+// lazySectionReader is a lazy reader for ELF sections, with reduced number of allocations.
+// It's oriented around an iterator pattern, where only one section is read (and allocated) at a time.
+type lazySectionReader struct {
+	reader              io.ReaderAt
+	err                 error
+	stringBuffer        []byte
+	sectionHeaderBuffer []byte
+	currentSection      *elfSection
+}
+
+// newLazySectionReader a new lazySectionReader instance.
+func newLazySectionReader(reader io.ReaderAt) *lazySectionReader {
+	return &lazySectionReader{
+		reader:         reader,
+		stringBuffer:   make([]byte, stringBufferSize),
+		currentSection: new(elfSection),
+	}
+}
+
+func (l *lazySectionReader) Err() error {
+	return l.err
+}
+
+// Iterate reads the ELF sections and yields them one by one. After iterating
+// through all sections, the error field l.err is set if an error occurred.
```

Comment:
maybe state that the code is copied from the debug/elf pkg

Reply:
actually add this to each function where it is true, or rather make a file-level general comment at the beginning